examples/ensemble/plot_clustering_on_random_trees_embedding_outputs_with_modified_distance_metrics.py

```diff
@@ -0,0 +1,177 @@
+
+# plot the heatmap 
+
+plt.subplot(1, 3, 1)
+plt.imshow(d_Nearest_Common_Ancestor)
+
+# use agglomerative clustering to classify the output
+
+from sklearn.cluster import AgglomerativeClustering
+from sklearn.metrics import adjusted_rand_score
+
+cluster = AgglomerativeClustering(n_clusters=3, affinity="precomputed", linkage="complete")
+predict_labels = cluster.fit_predict(d_Nearest_Common_Ancestor)
+score = adjusted_rand_score(labels, predict_labels)
+
+print("labels\n", labels)
+print("predict\n", predict_labels)
+print("Adjusted Rand Score:", score)
+
+# compute the second distance metrics, which is 
+# length of shortest path between leaf nodes
+# averaged over all estimators
+
+d = np.zeros([n_sample, n_sample])
+
+for k in range(n_estimator):
+    
+    tree = forest[k]
+    
+    path = tree.decision_path(samples).todense()
+    n_nodes = tree.tree_.node_count
+    l_children = tree.tree_.children_left
+    r_children = tree.tree_.children_right
+    
+    node_depth = np.zeros(n_nodes)
+    is_leaves = np.zeros(n_nodes)
+    stack = [(0, -1)]
+    
+    while len(stack) > 0:
+        node_id, parent_depth = stack.pop()
+        node_depth[node_id] = parent_depth + 1
+
+        if (l_children[node_id] != r_children[node_id]):
+            stack.append((l_children[node_id], parent_depth + 1))
+            stack.append((r_children[node_id], parent_depth + 1))
+        else:
+            is_leaves[node_id] = 1
+    
+    for i in range(n_sample):
+        for j in range(n_sample):
+            sample_ids = [i, j]
+            splitting_depth = (path[sample_ids].sum(axis=0) == len(sample_ids)).sum()
+            leaf_i = tree.apply(samples[i].reshape(1, -1))
+            leaf_j = tree.apply(samples[j].reshape(1, -1))
+            d[i, j] = d[i, j]+node_depth[leaf_i]+node_depth[leaf_j]-2*splitting_depth
+            
+d_shortest_path = d / n_estimator
+
+# plot the heatmap
+
+plt.subplot(1, 3, 2)
+plt.imshow(d_shortest_path)
+
+# use agglomerative clustering to classify the output
+
+cluster = AgglomerativeClustering(n_clusters=3, affinity="precomputed", linkage="complete")
+predict_labels = cluster.fit_predict(d_shortest_path)
+score = adjusted_rand_score(labels, predict_labels)
+
+print("labels\n", labels)
+print("predict\n", predict_labels)
+print("Adjusted Rand Score:", score)
+
+# compute the third distance metrics, which is 
+# the probability of two samples landing into
+# the same leaf node in an estimator
+
+Y = clf.transform(samples)
+print(Y.shape)
+prob = np.zeros([n_sample, n_sample])
+for i in range(n_sample):
+    for j in range(n_sample):
+        leaf_i = Y[i, :]
+        leaf_j = Y[j, :]
+        p = leaf_i.dot(leaf_j.transpose()).todense()
+        prob[i, j] = p[0, 0]
+
+prob = 1 - prob / Y.shape[1]
+        
+# plot the heatmap
+
+plt.subplot(1, 3, 3)
+plt.imshow(prob)
+
+# use agglomerative clustering to classify the output
+
+cluster = AgglomerativeClustering(n_clusters=3, affinity="precomputed", linkage="average")
```

Comment:
why does this one use a different linkage than the others?

Reply:
Turns out it gives better result.

Reply:
don't tell me, tell the next person who will read the notebook by updating the description/code accordingly. 

My point is mainly that since you haven't done any systematic parameter tuning (at least in the notebook that you are showing) then it could look like you are only picking certain results that tell the story you want to tell